### PR TITLE
Feature/kotlin migration

### PR DESCRIPTION
**Pull request by alvarez-fabian**
## Summary

Migrate the bootstap to use 100% kotlin.

Now it should be compatible and work without issues. Please keep in mind that I am testing it and it could have some errors.

I also added a couple of utils classes with function extensions that would simplify the code.

This uses `Android Kotlin Extensions` to bind the views, but we also keep `Butterknife` for special uses cases. (And because Wolmo continues using it). It'll be removed in the future.

The main consideration to have is the class `MiscModule` that provides a `BasePresenter<Any>`when you don't need a presenter in your classes.

Please let me know what do you think 😄

### CODE REVIEW

File: app/build.gradle [hunk 1]
```diff
@@ -85,18 +88,21 @@ jacocoAndroidUnitTestReport {
 }
 
 buildscript {
-    ext.support_library_version = '27.1.0'
+    ext.support_library_version = '27.1.1'
     ext.dagger_version = '2.13'
     ext.wolmo_version = '2.0.1'
-    ext.butterknife_version = '9.0.0-SNAPSHOT'
```
Comment (emalamela) [inline app/build.gradle]: Was this number lowered for any particular reasons?

File: app/src/main/java/ar/com/wolox/android/example/BaseConfiguration.kt [hunk 2]
```diff
@@ -0,0 +1,9 @@
+package ar.com.wolox.android.example
+
+internal open class BaseConfiguration {
```
Comment (emalamela) [inline app/src/main/java/ar/com/wolox/android/example/BaseConfiguration.kt]: Asymmetric blank lines for this class.

EDIT: These seems to be happening for most classes. Do you see any value in correcting this?

File: app/src/main/java/ar/com/wolox/android/example/BootstrapApplication.kt [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+import com.google.gson.FieldNamingPolicy
+import com.readystatesoftware.chuck.ChuckInterceptor
+import com.squareup.leakcanary.LeakCanary
+import dagger.android.AndroidInjector
+import okhttp3.logging.HttpLoggingInterceptor
+import okhttp3.logging.HttpLoggingInterceptor.Level
+
+class BootstrapApplication : WolmoApplication() {
+
+    override fun onInit() {
```
Comment (emalamela) [inline app/src/main/java/ar/com/wolox/android/example/BootstrapApplication.kt]: I would add a line inside the method saying something like "Initialize Application wide stuff here".

File: app/src/main/java/ar/com/wolox/android/example/ui/viewpager/random/RandomPresenter.kt [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+import ar.com.wolox.wolmo.core.presenter.BasePresenter
+import ar.com.wolox.wolmo.core.util.Logger
+import ar.com.wolox.wolmo.core.util.ToastFactory
+import java.util.*
+import javax.inject.Inject
+
+class RandomPresenter @Inject constructor(
+        private val userSession: UserSession,
+        private val toastFactory: ToastFactory
+) : BasePresenter<IRandomView>() {
```
Comment (emalamela) [inline app/src/main/java/ar/com/wolox/android/example/ui/viewpager/random/RandomPresenter.kt]: Does this indentation follow Kotlin style guide?

File: app/src/main/java/ar/com/wolox/android/example/di/module/MiscModule.kt [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+
+import ar.com.wolox.wolmo.core.presenter.BasePresenter
+import dagger.Module
+import dagger.Provides
+
+@Module
+class MiscModule {
+
+    @Provides
+    internal fun provideBasePresenter(): BasePresenter<Any> {
```
Comment (emalamela) [inline app/src/main/java/ar/com/wolox/android/example/di/module/MiscModule.kt]: Is this done through a `Provides` for any particular reason? I believe you can `@Inject constructor()` unless i'm missing something.

File: app/src/main/java/ar/com/wolox/android/example/ui/viewpager/ViewpagerActivity.kt [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+
+import ar.com.wolox.android.R
+import ar.com.wolox.android.example.ui.viewpager.fragment.ViewPagerFragment
+import ar.com.wolox.wolmo.core.activity.WolmoActivity
+import dagger.Lazy
+import javax.inject.Inject
+
+class ViewpagerActivity @Inject constructor() : WolmoActivity() {
+
+    // Lazy example, a lazy injection does not build the dependencies until #get() is called
```
Comment (emalamela) [inline app/src/main/java/ar/com/wolox/android/example/ui/viewpager/ViewpagerActivity.kt]: 👍

File: app/src/main/java/ar/com/wolox/android/example/ui/viewpager/request/RequestPresenter.kt [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+import ar.com.wolox.android.example.network.PostService
+import ar.com.wolox.android.example.utils.networkCallback
+import ar.com.wolox.wolmo.core.presenter.BasePresenter
+import ar.com.wolox.wolmo.networking.retrofit.RetrofitServices
+
+import javax.inject.Inject
+
+class RequestPresenter @Inject constructor(
+        private val mRetrofitServices: RetrofitServices
+) : BasePresenter<IRequestView>() {
```
Comment (emalamela) [inline app/src/main/java/ar/com/wolox/android/example/ui/viewpager/request/RequestPresenter.kt]: Same here.

**Comment by emalamela** [comment]
@alvarez-fabian Don't forget about us 😢.

File: app/build.gradle (hunk 1 above)
Comment (alvarez-fabian) [reply app/build.gradle]: It does not exist anymore :thinking:

File: app/src/main/java/ar/com/wolox/android/example/ui/viewpager/random/RandomPresenter.kt (hunk 4 above)
Comment (alvarez-fabian) [reply app/src/main/java/ar/com/wolox/android/example/ui/viewpager/random/RandomPresenter.kt]: Yep, is the one they [recommend](https://kotlinlang.org/docs/reference/coding-conventions.html#class-header-formatting).

File: app/src/main/java/ar/com/wolox/android/example/di/module/MiscModule.kt (hunk 5 above)
Comment (alvarez-fabian) [reply app/src/main/java/ar/com/wolox/android/example/di/module/MiscModule.kt]: Just to be an example of how to use it. (It was before in the Java code)

File: app/src/main/java/ar/com/wolox/android/example/ui/viewpager/request/RequestPresenter.kt (hunk 7 above)
Comment (alvarez-fabian) [reply app/src/main/java/ar/com/wolox/android/example/ui/viewpager/request/RequestPresenter.kt]: Yep, recommended.

**Comment by emalamela** [comment]
Fabi was removed from the Wolox organization and lost permission to modify this repo. He will fork it and then submit a PR to merge.

**Pull request closed without merging**